internal/scm/github/github.go

```diff
@@ -333,6 +342,76 @@ func (g *Github) getSearchRepositories(ctx context.Context, search string) ([]*g
 	return repos, nil
 }
 
+func (g *Github) getCodeSearchRepositories(ctx context.Context, search string) ([]*github.Repository, error) {
+	resultRepos := make(map[string]RepositoryReference)
+
+	i := 1
+	for {
+		rr, _, err := retry(ctx, func() ([]*github.CodeResult, *github.Response, error) {
+			rr, resp, err := g.ghClient.Search.Code(ctx, search, &github.SearchOptions{
+				ListOptions: github.ListOptions{
+					Page:    i,
+					PerPage: 100,
+				},
+			})
+
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if rr.GetIncompleteResults() {
+				// can occur when search times out on the server: for now, fail instead
+				// of handling the issue
+				return nil, nil, fmt.Errorf("search timed out on GitHub and was marked incomplete: try refining the search to return fewer results or be less complex")
+			}
+
+			if rr.GetTotal() > 1000 {
+				return nil, nil, fmt.Errorf("%d results for this search, but only the first 1000 results will be returned: try refining your search terms", rr.GetTotal())
+			}
+
+			return rr.CodeResults, resp, nil
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range rr {
+			repo := r.Repository
+
+			resultRepos[repo.GetFullName()] = RepositoryReference{
+				OwnerName: repo.GetOwner().GetLogin(),
+				Name:      repo.GetName(),
+			}
+		}
+
+		if len(rr) != 100 {
+			break
+		}
+		i++
+	}
+
+	// Code search does not return full details (like permissions). So for each
+	// repo discovered, we have to query it again.
+
+	repoNames := mapValues(resultRepos)
+	return g.getAllRepositories(ctx, repoNames)
+}
+
+func (g *Github) getAllRepositories(ctx context.Context, repoRefs []RepositoryReference) ([]*github.Repository, error) {
+	var repos []*github.Repository
+
+	for _, ref := range repoRefs {
```

Comment:
Nice to have (no need if you don't want to) would be to use the v4 (GraphQL) API to fetch all of these with one request.
That is true for --repo as well, but maybe even more here since I guess it will normally be more returns.

Reply:
Quite honestly I thought about this, but lazily shrank away from fooling with GraphQL. I'll have a look at it.

Perhaps if I added that in a separate PR, and unified the way this and `--repo` work?

Reply:
I've had a bit of a look into this, and I think I'll definitely hold this to a separate PR if possible.

I'm not sure that there's a direct equivalent of the `permissions` element returned by REST. There's a similar field (`viewerPermission`), but it [has some caveats](https://docs.github.com/en/graphql/reference/objects#repository).

Reply:
I'm going to resolve this conversation in favour of #414 and a later PR following from it.